### PR TITLE
Converted `CustomerInfo` and related types to use `Codable`

### PR DESCRIPTION
**Pull request by NachoSoto**
### Depends on:
- #1537
- #1540
- #1541
- #1543
- #1546
- #1547
- #1550
- #1551,
- #1565

### Changes

- Removed all custom deserialization code thanks to #1537, #1541, and #1543.
- The format of the customer info response is now very concise and only in one place (`CustomerInfoResponse`) instead of being passed through dictionaries and magic strings everywhere.
- `CustomerInfo` was combining the dictionaries inside of `"subscriptions"` and `"non_subscriptions"`, despite them having different formats. This is now made explicit through type conversions between the two, (see `CustomerInfoResponse.allTransactionsByProductId`).
- Removed all custom `CustomerInfo` errors since deserialization is automatic, and the underlying `Error` information will be provided thanks to `ErrorUtils.logDecodingError`.
- `CustomerInfo` deserialization always wraps errors into `ErrorCode.customerInfoError`, which is now also covered in a new test.
- Added a new snapshot test that covers `CustomerInfo` serialization by storing the result in a JSON file.

### Other changes:
- `CustomerInfo` equality is now automatic through the `CustomerInfoResponse` `Equatable` conformance.
- Simplified `schemaVersion` handling
- All tests still use `CustomerInfo(data: [String: Any])`, but that method is only visible for tests now, and it uses the underlying `Decodable` implementation (see `CustomerInfo+TestExtensions.swift`).
- Added extra tests for `CustomerInfoResponse` that use a fixture `CustomerInfo.json`, which is easier to maintain than having JSON in code.
- `RawDataContainer` implementation is changed in #1565.

### TODO:
- [x] Fix all tests
- [x] Update `RawValueContainer`
- [x] Consider recovering `CustomerInfoError`
- [x] Look at all previously logged errors
- [x] Finish `CustomerInfoResponseTests`
- [x] Handle all TODOs
- [x] Make `RawDataContainer` actually store all the original data: #1565

### CODE REVIEW

**Comment by NachoSoto** [comment]
Down to just 7 failing tests 🎉 almost there!

File: Sources/CodableExtensions/EntitlementData+Extensions.swift [hunk 1]
```diff
@@ -1,30 +0,0 @@
...
-//
-//      https://opensource.org/licenses/MIT
-//
-//  EntitlementData+Extensions.swift
-//
-//  Created by Juanpe Catalán on 27/8/21.
-
-import Foundation
-
-extension EntitlementInfo.EntitlementData {
```
Comment (NachoSoto) [inline Sources/CodableExtensions/EntitlementData+Extensions.swift:16]: This custom deserialization is no longer needed

File: Sources/CodableExtensions/EntitlementProductData+Extensions.swift [hunk 2]
```diff
@@ -1,38 +0,0 @@
...
-//
-//      https://opensource.org/licenses/MIT
-//
-//  EntitlementProductData+Extensions.swift
-//
-//  Created by Juanpe Catalán on 27/8/21.
-
-import Foundation
-
-extension EntitlementInfo.ProductData {
```
Comment (NachoSoto) [inline Sources/CodableExtensions/EntitlementProductData+Extensions.swift:16]: This was removed in #1540.

File: Sources/Identity/CustomerInfo.swift [hunk 3]
```diff
@@ -87,15 +86,6 @@ import Foundation
...
     @objc public let originalApplicationVersion: String?
 
-    /**
-     * The underlying data for this `CustomerInfo`.
-     *
-     * - Note: the content and format of this data isn’t documented and is subject to change.
-     *         it's only meant for debugging purposes or for getting access to future data
-     *         without updating the SDK.
-     */
-    @objc public let rawData: [String: Any]
```
Comment (NachoSoto) [inline Sources/Identity/CustomerInfo.swift:97]: This is now provided automatically through `RawDataContainer`.

File: Sources/Identity/CustomerInfo.swift [hunk 4]
```diff
@@ -129,13 +119,12 @@ import Foundation
             return false
         }
 
-        return NSDictionary(dictionary: self.jsonObjectWithNoDate)
```
Comment (NachoSoto) [inline Sources/Identity/CustomerInfo.swift:132]: Much simpler comparison! `CustomerInfoResponse` `Equatable` implementation ignores the date.

File: Sources/Identity/CustomerInfo.swift [hunk 5]
```diff
@@ -164,263 +153,168 @@ import Foundation
...
-                json: customerJSON.stringRepresentation
-            )
-        }
+    /// Initializes a `CustomerInfo` creating a copy.
+    convenience init(customerInfo: CustomerInfo) {
+        self.init(data: customerInfo.data)
     }
 
-    static let currentSchemaVersion = "2"
+    fileprivate init(data: Data) {
```
Comment (NachoSoto) [inline Sources/Identity/CustomerInfo.swift:170]: This method now stores each of the properties from the already decoded `Data`.

File: Sources/Identity/CustomerInfo.swift [hunk 6]
```diff
@@ -164,263 +153,168 @@ import Foundation
...
-            let latestNonSubscriptionTransactionsByProductId = [String: [String: Any]](
-                uniqueKeysWithValues: nonSubscriptionsByProductId.map { productId, transactionsArray in
-                    (productId, transactionsArray.last ?? [:])
-                })
-
-            self.allTransactionsByProductId = latestNonSubscriptionTransactionsByProductId
-                .merging(subscriptionTransactionsByProductId, strategy: .keepOriginalValue)
-
-            self.allPurchases = latestNonSubscriptionTransactionsByProductId
-                .merging(subscriptionTransactionsByProductId) { (current, _) in current }
```
Comment (NachoSoto) [inline Sources/Identity/CustomerInfo.swift:338]: This was super hacky and hard to follow. It was combining the dictionaries inside of "subscriptions" and "non_subscriptions", despite them having different formats. This is now made explicit through type conversions between the two.

File: Sources/Networking/Responses/CustomerInfoResponse.swift [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                     transactionIdentifier: nil,
+                     store: self.store,
+                     isSandbox: self.isSandbox)
+    }
+
+}
+
+extension CustomerInfoResponse.Subscriber {
+
+    var allTransactionsByProductId: [String: CustomerInfoResponse.Transaction] {
```
Comment (NachoSoto) [inline Sources/Networking/Responses/CustomerInfoResponse.swift:177]: `CustomerInfo` was combining the dictionaries inside of "subscriptions" and "non_subscriptions", despite them having different formats. This is now made explicit through type conversions between the two.

**Review by NachoSoto — COMMENTED** [review]
This is ready for review now!

File: Sources/Identity/CustomerInfo.swift (hunk 6 above)
Comment (joshdholtz) [reply Sources/Identity/CustomerInfo.swift:338]: OMG, the new version is so ❤️ I was in here debugging something a while back and it was hard to follow. Great work!

File: Sources/Identity/CustomerInfoManager.swift [hunk 8]
```diff
@@ -120,45 +120,26 @@ class CustomerInfoManager {
...
-                    Logger.error("Error loading customer info from cache: \(error)")
-                }
-                return nil
-            }
-
-            if let schema = info.schemaVersion, schema == CustomerInfo.currentSchemaVersion {
+            if info.isInCurrentSchemaVersion {
                 return info
+            } else {
+                return nil
```
Comment (joshdholtz) [inline Sources/Identity/CustomerInfoManager.swift:128]: Do we need to log anything here for this error? 🤔

File: Sources/Misc/RawDataContainer.swift [hunk 9]
```diff
@@ -11,14 +11,33 @@
...
+    /// to future data while using an older version of the SDK.
+    @available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.2, *)
+    public var rawData: [String: Any] {
+        do {
+            return try self.underlyingData.asDictionary()
+        } catch {
+            Logger.warn(Strings.codable.encoding_error(error))
+            return [:]
+        }
+    }
```
Comment (joshdholtz) [inline Sources/Misc/RawDataContainer.swift:41]: Will this `rawData` not contain any data that was stripped away by the new lossy array and dictionary types? So like.. this won't show the raw response that this customer info was created from, right? It instead is just a dictionary representation of this object?

I'm not sure if that is a change from what this previously did but just wanted to understand and call that out just in case 😇

File: Sources/Networking/Responses/CustomerInfoResponse.swift [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+//
+//      https://opensource.org/licenses/MIT
+//
+//  CustomerInfoResponse.swift
+//
+//  Created by Nacho Soto on 4/12/22.
+
+import Foundation
+
+struct CustomerInfoResponse {
```
Comment (joshdholtz) [inline Sources/Networking/Responses/CustomerInfoResponse.swift:16]: This whole file is soooooo 💯

File: Sources/Networking/Responses/CustomerInfoResponse.swift [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+    var allTransactionsByProductId: [String: CustomerInfoResponse.Transaction] {
+        return self.allSubscriptionsByProductId.mapValues { $0.asTransaction }
+    }
+
+    var allSubscriptionsByProductId: [String: CustomerInfoResponse.Subscription] {
+        let subscriptions = self.subscriptions
+        let latestNonSubscriptionTransactionsByProductId = self.nonSubscriptions
+            .compactMapValues { $0.last }
+            .mapValues { $0.asSubscription }
```
Comment (joshdholtz) [inline Sources/Networking/Responses/CustomerInfoResponse.swift:185]: Why are non-subscriptions being included in `allSubscriptions` 🤔

**Review by joshdholtz — COMMENTED** [review]
Leaving some early comments on my review while I continue reviewing 😊

File: Sources/Identity/CustomerInfoManager.swift (hunk 8 above)
Comment (NachoSoto) [reply Sources/Identity/CustomerInfoManager.swift:128]: We weren't logging anything before 🤷🏻‍♂️

File: Sources/Networking/Responses/CustomerInfoResponse.swift (hunk 11 above)
Comment (NachoSoto) [reply Sources/Networking/Responses/CustomerInfoResponse.swift:185]: This was like this before:
```swift
self.allPurchases = latestNonSubscriptionTransactionsByProductId
    .merging(subscriptionTransactionsByProductId) { (current, _) in current }
```
Maybe I should change it to `allPurchasesByProductId` instead?

File: Sources/Misc/RawDataContainer.swift (hunk 9 above)
Comment (NachoSoto) [reply Sources/Misc/RawDataContainer.swift:41]: Oh shoot that's a great catch! Let me look into how we could support storing the original data.

File: Sources/Networking/Responses/CustomerInfoResponse.swift (hunk 11 above)
Comment (joshdholtz) [reply Sources/Networking/Responses/CustomerInfoResponse.swift:185]: I _think_ that would make sense 🤷‍♂️

File: Sources/Identity/CustomerInfoManager.swift (hunk 8 above)
Comment (joshdholtz) [reply Sources/Identity/CustomerInfoManager.swift:128]: Probably fine then!

File: Sources/Networking/Responses/CustomerInfoResponse.swift (hunk 11 above)
Comment (NachoSoto) [reply Sources/Networking/Responses/CustomerInfoResponse.swift:185]: DOne.

File: Sources/Identity/CustomerInfo.swift [hunk 12]
```diff
@@ -164,263 +153,170 @@ import Foundation
...
 
+extension CustomerInfo: HTTPResponseBody {}
+
+// MARK: - Private
+
 private extension CustomerInfo {
 
-    static func createSubscriberDataError(_ error: Error, subscriberDictionary: [String: Any]) -> Error {
-        Logger.error(Strings.customerInfo.cant_instantiate_from_json_object(jsonObject: subscriberDictionary))
+    /// The actual contents of a `CustomerInfo`: the response with the associated version.
```
Comment (aboedo) [inline Sources/Identity/CustomerInfo.swift:255]: `CustomerInfo` here should be in \`\`double backticks\`\` so you can click on it to navigate

Comment (aboedo) [inline Sources/Identity/CustomerInfo.swift:255]: `CustomerInfo` here should be in `\`\double backticks \`\` so you can click on it to navigate

File: Sources/Identity/CustomerInfo.swift [hunk 13]
```diff
@@ -164,263 +153,170 @@ import Foundation
...
+    struct Data {
```
Comment (aboedo) [inline Sources/Identity/CustomerInfo.swift:256]: this confused me when reading the decoding code. I initially thought we'd added `response` and `schemaVersion` as extra properties in an extension of Foundation's Data. 
I think this approach makes sense, but perhaps we can call it something other than `Data` so there's no confusing it with a standard type?

File: Sources/Identity/CustomerInfo.swift [hunk 14]
```diff
@@ -164,263 +153,170 @@ import Foundation
...
 }
 
-// Fixme: delete when `CustomerInfo` is `Decodable`
- extension CustomerInfo: HTTPResponseBody {
+private extension CustomerInfo {
+
+    func activeKeys(dates: [String: Date?]) -> Set<String> {
+        return Set(
+            dates
+                .lazy
```
Comment (aboedo) [inline Sources/Identity/CustomerInfo.swift:303]: why is `lazy` needed here?

File: Sources/Misc/RawDataContainer.swift [hunk 15]
```diff
@@ -11,14 +11,33 @@
...
 //  Created by Nacho Soto on 11/16/21.
 
-/// A type which exposes its underlying raw data, for debugging purposes or for getting access
+/// A type which exposes its underlying content for debugging purposes or for getting access
 /// to future data while using an older version of the SDK.
 public protocol RawDataContainer {
 
-    /// The type of the underlying raw data
-    associatedtype Content
+    /// The type of the `underlyingData` for this type.
```
Comment (aboedo) [inline Sources/Misc/RawDataContainer.swift:18]: I think you get automatic linking to the symbol if you do \`\`RawDataContainer/underlyingData\`\` here

File: Sources/Misc/RawDataContainer.swift [hunk 16]
```diff
@@ -11,14 +11,33 @@
...
-    /// The underlying data.
-    var rawData: Content { get }
+    /// The underlying content for debugging purposes or for getting access
+    /// to future data while using an older version of the SDK.
+    @available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.2, *)
+    var underlyingData: Content { get }
+
+}
+
+/// Default implementation that encodes the `underlyingData` for public use.
```
Comment (aboedo) [inline Sources/Misc/RawDataContainer.swift:28]: same thinking regarding auto linking to the symbol

File: Tests/UnitTests/Networking/Responses/CustomerInfoDecodingTests.swift [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        try super.setUpWithError()
+
+        self.customerInfo = try self.decodeFixture("CustomerInfo")
+    }
+
+    func testSchemaVersion() {
+        expect(self.customerInfo.schemaVersion) == "4"
+    }
+
+    func testResponseDataIsCorrect() throws {
```
Comment (aboedo) [inline Tests/UnitTests/Networking/Responses/CustomerInfoDecodingTests.swift:35]: maybe we can spilt this up into multiple cases? the response is likely to grow over time, and the test case is already pretty big

Comment (aboedo) [inline Tests/UnitTests/Networking/Responses/CustomerInfoDecodingTests.swift:35]: maybe we can spilt this up into multiple cases? the response is likely to grow over time, and the test case is already pretty big

**Review by aboedo — COMMENTED** [review]
This looks amazing!! It's great to see the CustomerInfo decoding code go away, that was a pain to read. 

I left a few comments

File: Sources/Identity/CustomerInfo.swift (hunk 13 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Identity/CustomerInfo.swift:256]: Good point. The whole type is `CustomerInfo.Data`, maybe I can use that whole type for clarity?

[the remark this answers is not in the document]
Comment (aboedo) [reply Sources/Identity/CustomerInfo.swift:256]: I think that works, although the reason I suggested renaming was so that in the future we don't use it by calling it `Data` and have this conversation again

[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Identity/CustomerInfo.swift:256]: That's fair. Maybe `CustomerInfo.Contents`?

File: Sources/Identity/CustomerInfo.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Identity/CustomerInfo.swift:303]: Avoids creating intermediate 3 intermediate collections, one for filtering, one for mapping, and then another copy for the set. This instead initializes the set going through each element only once.

File: Sources/Misc/RawDataContainer.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Misc/RawDataContainer.swift:18]: Oh yeah thanks! Changed.

[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Misc/RawDataContainer.swift:18]: Actually this is going away in #1565 (I put it separate because this was already big enough).

File: Sources/Purchasing/EntitlementInfos.swift [hunk 18]
```diff
@@ -43,78 +43,55 @@ import Foundation
...
+                }
+
+                return (
+                    identifier,
+                    EntitlementInfo(identifier: identifier,
+                                    entitlement: entitlement,
+                                    subscription: subscription,
+                                    requestDate: requestDate)
+                )
+            }
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Purchasing/EntitlementInfos.swift:91]: Yeah I guess it's personal preference, I'm happy to change the data manipulation method back.
What do you mean by "flat" though?

File: Tests/UnitTests/Networking/Responses/CustomerInfoDecodingTests.swift (hunk 17 above)
Comment (NachoSoto) [reply Tests/UnitTests/Networking/Responses/CustomerInfoDecodingTests.swift:35]: Good idea.

Comment (NachoSoto) [reply Tests/UnitTests/Networking/Responses/CustomerInfoDecodingTests.swift:35]: Done. I changed the `setUp` to `class` so it only happens one for all tests.

Comment (NachoSoto) [reply Tests/UnitTests/Networking/Responses/CustomerInfoDecodingTests.swift:35]: Done.

File: Sources/Misc/RawDataContainer.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Misc/RawDataContainer.swift]: Never mind, the type is still there, so I updated the docstring 👍🏻

File: Sources/Identity/CustomerInfo.swift (hunk 14 above)
Comment (aboedo) [reply Sources/Identity/CustomerInfo.swift:303]: it seems like gh posted my comments once per time I edited them before posting the review? I find a different but in gh for every PR over 700 lines.

File: Sources/Identity/CustomerInfo.swift (hunk 13 above)
Comment (NachoSoto) [reply Sources/Identity/CustomerInfo.swift:256]: Changed!

[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Identity/CustomerInfo.swift:256]: Changed!

File: Sources/Misc/RawDataContainer.swift (hunk 9 above)
Comment (NachoSoto) [reply Sources/Misc/RawDataContainer.swift:41]: Forgot to update this here, this is fixed by #1565.

File: Sources/Purchasing/EntitlementInfos.swift (hunk 18 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Purchasing/EntitlementInfos.swift:91]: > well, not really flat, it's just that the new implementation does the filtering and matching between entitlementInfos and purchases as an operation within the constructor of the entitlementsByKey dictionary, which seems like a lot to do at once.

Yeah, one is imperative and this is functional. It's kind of a philosophical argument more than anything :P 

> It also took me a bit to realize that the entitlements in the signature and the one in declared in the method are different objects

Good point, I'll get rid of the unnecessary variable.

**Review by aboedo — APPROVED** [review]

File: Sources/Misc/RawDataContainer.swift [hunk 19]
```diff
@@ -11,14 +11,33 @@
...
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Misc/RawDataContainer.swift:28]: I changed this in #1565 already :)
